### DISCUSSION TITLE
SignIn using Email Provider server side

### DISCUSSION THREAD

**Discussion by pbteja1998** [nextauthjs/next-auth#1627] [Help] [answered]
**Your question**
Is it possible to `signin` the user using the email provider, server-side?

**What are you trying to do**
I have a stripe webhook setup, through which I will get the email of the new user. Now, I want to create an account with this email, and also send the verification email (to sign in).

**What methods I have tried**
1. I tried using the `signIn` method – `signIn('email', {redirect: false, email: customer.email})`, but it throws an error saying `window is not defined`.
2. I tried directly sending the `POST` request directly with the `csrfToken` and `email`, but it just returns a HTML text.
```javascript
// This code is being executed server-side in an API route.

const csrfToken = await getCsrfToken({ req })

// and then I am using the above `csrfToken` in the `fetch` request.
fetch(
    'http://localhost:3000/api/auth/signin/email',
    {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ csrfToken, email: customer.email }),
    }
  )
```

**Reproduction**
// Added above

**Feedback**
*Documentation refers to searching through [online documentation](https://next-auth.js.org), code comments and issue history. The example project refers to [next-auth-example](https://github.com/iaincollins/next-auth-example).*

* [ ] Found the documentation helpful
* [ ] Found documentation but was incomplete
* [ ] Could not find relevant documentation
* [ ] Found the example project helpful
* [ ] Did not find the example project helpful

**Comment by pbteja1998** [comment]
Updated to the latest version of `next-auth`. Now I am getting different response, but still no email is being sent.

```javascript
const response = await signIn('email', {
  redirect: false,
  email: customer.email,
  callbackUrl: '',
})
```

The above code is returning the following response.
```js
{
  error: null,
  status: 200,
  ok: true,
  url: 'http://localhost:3000/api/auth/signin?csrf=true'
}
```

**Comment by balazsorban44** [comment]
hmm, the response you are providing suggests that it doesn't accept your csrfToken for some reason. if I had to guess, the page you are visiting for the first time must not be the one that is making this request, because the CSRF cookie hasn't been set yet. try loading the page from /, and then do a clientside navigation to the lage that is using getServerSideProps. does it work then? 

the signIn method not working server-side is expected, as it is meant to be used client-side.

**Reply by pbteja1998** [reply]
Is there a way to achieve the workflow of `signIn` method server side. May be it would help if I explained my usecase in more detail. 

---

I have a Stripe subscription pricing page. Users(not logged in) will click on the link, start the Stripe Checkout session. They will then be redirected to Stripe hosted checkout page, enters their email, their card details, and pays for the subscription.

Meanwhile, I have an api route in NextJS app that listens to different events generated by Stripe. One of those is `subscription.created` event.

This event is emitted when the user successfully completed the subscription flow. 

Now, what I wanted to do was, whenever I receive this event, I also get the email that the user used while paying for the subscription, through the event payload. So, I want to use this email, and send the sign in link to the user. 

Please note that, this entire flow happens in an api route, since that is where I set up my webhook and am listening to Stripe events. 

Now, can you please tell me if there is a way to achieve what I wanted to do with `next-auth`?

**Reply by pbteja1998** [reply]
> try loading the page from /, and then do a clientside navigation to the lage that is using getServerSideProps. does it work then?

I won't ever be visiting the page where I am making the request, since it's an api route that acts as a webhook. It just listens to the events emitted by the Stripe, and acknowledges them by sending status 200 back.

**Comment by willllhallll** [comment]
We have been looking into too; the ability to send out user invites programmatically would be a great feature. It also looks like this has been brought up a few times too (#141, #703, #1229). 

With regards to the CSRF request issues, I think the best explanation I have comes down to no cookies = no CSRF. In a RESTful setting the CSRF cookie is unset and so call calls to `csrfTokenHandler` will return a new token every time: 

https://github.com/nextauthjs/next-auth/blob/76df2b5e702be2084674c81e098d989c2ec39ed9/src/server/lib/csrf-token-handler.js#L34-L40

So when you POST to `/api/auth/signin/email/`, the CSRF verification fails. 

https://github.com/nextauthjs/next-auth/blob/073da60c3d6c06cfa0ee0918cba0d65b55faf898/src/server/index.js#L70

https://github.com/nextauthjs/next-auth/blob/073da60c3d6c06cfa0ee0918cba0d65b55faf898/src/server/index.js#L200-L206

(_I have no idea if any of this is correct, just my first guess :) Calling all wise people to verify this is the case!_)

I haven't yet come up with an interim solution, and would be excited to help contribute to such an 'invites' feature, but maybe in the meantime you could dip into the `next-auth` source and pull out `dist/server/lib/signin/email` to bypass all CSRF stuff. Construction the required provider and options might be a pain, and it might be important to pass a `state` parameter throughout your Stripe flow to verify the origin of requests to your server-only email-sending endpoint, but I think it could be done. There could also be many reasons why this is a terrible idea... So... 

But looking forward, an invite flow of some kind would be sweet. Would love to help out and contribute to something like that.

**Reply by pbteja1998** [reply]
Thanks very much for the input and the similar issues. The following comment seems promising, will try to do as they suggested and see if it works.

https://github.com/nextauthjs/next-auth/issues/703#issuecomment-697168310

**Comment by pbteja1998** [comment, marked as answer]
I was able to solve this with the help of https://github.com/nextauthjs/next-auth/issues/703#issuecomment-812070912 by @guiaramos (🙏)

This is what my code looks like:
```javascript
import cookie from 'cookie'
// getURL() returns http://localhost:3000 locally


let Cookie
const response = await fetch(`${getURL()}/api/auth/csrf`).then(
  (res) => {
    const parsedCookie = cookie.parse(res.headers.get('set-cookie'))
    delete parsedCookie.Path
    delete parsedCookie.SameSite
    Cookie = Object.entries(parsedCookie)
      .map(([key, val]) => cookie.serialize(key, val))
      .join('; ')
    return res.json()
  }
)
const csrfToken = response.csrfToken
const fetchOptions = {
  method: 'POST',
  headers: {
    'Content-Type': 'application/x-www-form-urlencoded',
    Cookie,
  },
  body: new URLSearchParams({
    email: customer.email,
    callbakUrl: getURL(),
    redirect: 'false',
    csrfToken,
    json: 'true',
  }),
}

const res = await fetch(`${getURL()}/api/auth/signin/email`, fetchOptions).then((res) => res.json())
```

This is successfully sending a verification request email to the corresponding email.

**Reply by willllhallll** [reply]
That's great!

I was wondering if you had implemented any way of checking that calls to your new endpoint are legitimately from Stripe, rather than just any `POST` request? Like some kind of `state`? 

Otherwise, I would imagine it could be easy for people to get arbitrary email invites sent out (and access to the app without them having followed the correct subscription flow) , simply by making requests this endpoint. 

Would love to hear your thoughts. Cheers.

**Reply by pbteja1998** [reply]
Yes, Stripe has that mechanism built in. It sends `Stripe-Signature` header. You would be able to verify this signature using the `webhook secret` that is present in the dashboard. Only after everything is verified, we will look at the actual event and act on it. So, there is no way for others to send requests directly to this endpoint, and then get an invite.

**Reply by pbteja1998** [reply]
This is how the code roughly looks like:
```javascript
const buf = await buffer(req)
const sig = req.headers['stripe-signature']
const webhookSecret = env.STRIPE_WEBHOOK_SECRET
let event: Stripe.Event
try {
  event = stripe.webhooks.constructEvent(buf.toString(), sig, webhookSecret)
} catch (error) {
  console.log(`❌ Error message: ${error.message}`)
  res.status(400).send(`Webhook Error: ${error.message}`)
  return
}
```

**Reply by willllhallll** [reply]
That is super neat, thanks for sharing. I had no idea about `signature-headers` before, this is great!

To share what I have found if others stumble across the same question; it seems like the convention is to make such webhook routes as this exempt from `CSRF` protection anyway:

- CSRF Protection: https://stripe.com/docs/webhooks/best-practices#csrf-protection

Many services use these `signature-headers` instead with a secret defined and shared well ahead of deployment:

- Stripe: https://stripe.com/docs/webhooks/signatures
- Paddle: https://developer.paddle.com/webhook-reference/verifying-webhooks

**Reply by gaelmotte** [reply]
Hey ! 
Based on your idea, i propose we do not fetch, instead, we call the handler by handcrafting NextRequests.

Suggestion comes from the fact an app deployed on vercel may not call itself in preview stages, due to vercel authenticating users accessing the product.
I'm pretty sure other PaaS providerd may have the same behavior.

```typescript
import cookie from 'cookie'

import {GET, POST} from "./auth"
import { NextRequest } from 'next/server';


function getURL(){
    if (!process.env.NEXTAUTH_URL) throw new Error("NEXTAUTH_URL is required");
    return process.env.NEXTAUTH_URL;
}

export default async function(email:string){

    const requestCSRF: NextRequest = new NextRequest(`${getURL()}/api/auth/csrf`);
    let Cookie: string = "";
    const response = await GET(requestCSRF).then(
        (res) => {
          const csrfCookie = res.headers.get('set-cookie');
          if(!csrfCookie) throw new Error("Failed to get CSRF cookies");
          const parsedCookie = cookie.parse(csrfCookie)
          delete parsedCookie.Path
          delete parsedCookie.SameSite
          Cookie = Object.entries(parsedCookie)
            .map(([key, val]) => cookie.serialize(key, val))
            .join('; ')
          return res.json()
        }
      );

    const csrfToken = response.csrfToken

    const requestSignin: NextRequest = new NextRequest(`${getURL()}/api/auth/signin/nodemailer`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
          Cookie,
        },
        body: new URLSearchParams({
          email: email,
          callbakUrl: getURL(),
          redirect: 'false',
          csrfToken,
          json: 'true',
        }),
      });
    

    
    await POST(requestSignin);
    
    ```

**Comment by CoolTechYT** [comment]
Hello! 

I am writing because I am attempting to do something very similar with the credentials provider. Here is my code:

```.js
import qs from 'qs';
import cookie from 'cookie';
import axios from 'axios';

const csrfData = async () => await axios({
    method: 'GET',
    url: `${process.env.NEXTAUTH_URL}api/auth/csrf`
});

export const getSignIn = async (email, password) => {
    const { data, headers } = await csrfData();

    const params = {
        email,
        password,
        redirect: 'false',
        csrfToken: String(data.csrfToken),
        json: 'true'
    };

    const body = qs.stringify(params);

    const parsedCookie = cookie.parse(headers['set-cookie'].join('; '));
    delete parsedCookie.Path;
    delete parsedCookie.SameSite;

    const Cookie = Object.entries(parsedCookie)
        .map(([key, val]) => cookie.serialize(key, val))
        .join('; ');

    const options = {
        method: 'POST',
        url: `${process.env.NEXTAUTH_URL}api/auth/callback/credentials?`,
        headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
            Cookie
        },
        data: body
    };

    return await axios(options);
}
```

What I am noticing is, it's successfully authenticating using `authorize` within the `CredentialsProvider`, and `api/auth/callback/credentials` returns 200 / 401 based on successful/unsuccessful credentials.

However, I'm not sure what to do next? The request returns 200, but it doesn't seem like the session is being set from here. Any help is appreciated.

**Reply by CoolTechYT** [reply]
Seems like I may have found out what to do with this.

I'm consuming `getSignIn` on a Next.js API route. Using it like this seems to work:

```.js
const signInResponse = await getSignIn(email, password);
res.setHeader("set-cookie", signInResponse.headers['set-cookie']);
return res.redirect(301, '/some-protected-route');
```

Basically, just need to set the response 'set-cookie' header using the 'set-cookie' header from the getSignIn.
